src/compiler/binder.ts

```diff
@@ -1689,9 +1689,15 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
         const clauses = node.clauses;
         const isNarrowingSwitch = node.parent.expression.kind === SyntaxKind.TrueKeyword || isNarrowingExpression(node.parent.expression);
         let fallthroughFlow = unreachableFlow;
+        let afterPreviousClauseFlow: FlowNode | undefined;

```

Comment:
This can be removed with a patch like this
<details>
<summary>git diff</summary>

```diff
diff --git a/src/compiler/binder.ts b/src/compiler/binder.ts
index b3840cc157..b26af72dcc 100644
--- a/src/compiler/binder.ts
+++ b/src/compiler/binder.ts
@@ -1689,14 +1689,12 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
         const clauses = node.clauses;
         const isNarrowingSwitch = node.parent.expression.kind === SyntaxKind.TrueKeyword || isNarrowingExpression(node.parent.expression);
         let fallthroughFlow = unreachableFlow;
-        let afterPreviousClauseFlow: FlowNode | undefined;
 
         for (let i = 0; i < clauses.length; i++) {
             const clauseStart = i;
             while (!clauses[i].statements.length && i + 1 < clauses.length) {
-                if (afterPreviousClauseFlow && afterPreviousClauseFlow === unreachableFlow) {
+                if (fallthroughFlow === unreachableFlow) {
                     currentFlow = preSwitchCaseFlow!;
-                    afterPreviousClauseFlow = undefined;
                 }
                 bind(clauses[i]);
                 i++;
@@ -1707,7 +1705,7 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
             currentFlow = finishFlowLabel(preCaseLabel);
             const clause = clauses[i];
             bind(clause);
-            fallthroughFlow = afterPreviousClauseFlow = currentFlow;
+            fallthroughFlow = currentFlow;
             if (!(currentFlow.flags & FlowFlags.Unreachable) && i !== clauses.length - 1 && options.noFallthroughCasesInSwitch) {
                 clause.fallthroughFlowNode = currentFlow;
             }
```
</details>

but I'm not yet 100% sure if I can repeatedly "reset" the `currentFlow` to `preSwitchCaseFlow` for consecutive statementless clauses.

Reply:
I came to the conclusion that this patch should be fine. However, I found a new small bug:
```ts
function test(
  foo:
    | { kind: "a"; prop: string }
    | { kind: "b"; prop: number }
    | { kind: "c"; prop: boolean },
  bar?: {
    type: "c";
  },
) {
  if (!bar) {
    return;
  }

  switch (foo.kind) {
    case "a":
      return;
    case ((bar = undefined), "b"):
    case bar.type: // should error
      return;
  }
}
```

It's a separate issue since it doesn't work on `main` either.